iwyu.cc

```diff
@@ -1518,15 +1512,9 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return retval;
   }
 
-  set<const Type*> GetCallerResponsibleTypesForFnReturn(
-      const FunctionDecl* decl) const {
-    set<const Type*> retval;
-    const Type* return_type = Desugar(decl->getReturnType().getTypePtr());
-    if (CodeAuthorWantsJustAForwardDeclare(return_type, GetLocation(decl))) {
-      retval.insert(return_type);
-      // TODO(csilvers): include template type-args if appropriate.
-    }
-    return retval;
+  set<const Type*> GetProvidedTypesForFnReturn(const FunctionDecl* decl) const {
+    const Type* return_type = decl->getReturnType().getTypePtr();
```

Comment:
This used to Desugar, but no longer does -- that's intentional to make sure we pick up as much sugar as possible, right?

Reply:
No, desugaring here is rather redundant now. `TypeEnumeratorWithoutSubstituted` collects canonical types. Maybe, avoiding of desugaring helps with template functions returning template parameter type, but I'm not sure.

Reply:
> Maybe, avoiding of desugaring helps with template functions returning template parameter type, but I'm not sure.

I forgot it:
> No explicit return type reporting for instantiated function templates
because `InstantiatedTemplateVisitor` already does that taking into
account template argument underlying type provision status.

Reply:
Makes sense, thanks!